### PR TITLE
PR for merging work done in the context of INDIGO

### PR DESCRIPTION
**Pull request by gwarf**
Big/heavy PR for merging work done in the contect of INDIGO.
- Add --template-extension parameter allowing to select template file extension
- Add test data for docker images
- Add a (temporary probably) indigoon provider, returning images with all their metadata and templates as required for INDIGO
- OpenStack provider now returns all the available metadata
- Use Mako templating system (existing templates have been ported)
- Rework/complement some *BDII classes tests
- Rename cloud_bdii package to cloud_info
- Change --yaml-file default: from /etc/cloud-info-provider/bdii.yaml to /etc/cloud-info-provider/static.yaml
- Drop recommends bdii from debian package, as it is for RHEL
- Add travis build status badge
- Enable python 3.4 tests with Travis
- Update documentation
- Fix python-novaclient deprecation warning

### CODE REVIEW

**Comment by spinto** [comment]
I am out of the office until 10/08/2016.

For urgent matter, please contact the TEP Core Team (tepcoreteam@esa.int).


Note: This is an automated response to your message
"[EGI-FCTF/cloud-bdii-provider] PR for merging work done in the context of
INDIGO (#30)" sent on 8/9/2016 5:03:29 PM.

This is the only notification you will receive while this person is away.


This message and any attachments are intended for the use of the addressee or addressees only.
The unauthorised disclosure, use, dissemination or copying (either in whole or in part) of its
content is not permitted.
If you received this message in error, please notify the sender and delete it from your system.
Emails can be altered and their integrity cannot be guaranteed by the sender.

Please consider the environment before printing this email.

**Comment by gwarf** [comment]
Make ComputeBDII class use only one template and loop over content inside the template.

File: README.md [hunk 1]
```diff
@@ -25,17 +34,31 @@ Use the appropriate repos for your distribution and install using the usual tool
...
+sudo mkdir /etc/cloud-info-provider
+sudo chgrp you_user /etc/cloud-info-provider
+sudo chmod g+rwx /etc/cloud-info-provider
+```
+
 ```
-git clone https://github.com/EGI-FCTF/BDIIscripts
-cd BDIIscripts 
+git clone https://github.com/EGI-FCTF/cloud-bdii-provider
+cd cloud-bdii-provider
```
Comment (alvarolopez) [inline README.md]: This should be changed to cloud-info-provider, right? When the PR is merged, we can rename the repository (in Github) itself.

**Comment by alvarolopez** [comment]
Hi @gwarf a couple of comments:

- Dependencies: could you add "prb" and "python-novaclient" to the requirements list?

- When I try to run the script from this PR, I get the following error (probably because I am using a too new novaclient):

```
/tmp/bdii/local/lib/python2.7/site-packages/novaclient/client.py:831: UserWarning: 'get_client_class' is deprecated. Please use `novaclient.client.Client` instead.
  warnings.warn(_LW("'get_client_class' is deprecated. "
Traceback (most recent call last):
  File "/tmp/bdii/bin/cloud-info-provider-service", line 10, in <module>
    sys.exit(main())
  File "/tmp/bdii/cloud-bdii-provider/cloud_bdii/core.py", line 218, in main
    bdii = cls_(opts)
  File "/tmp/bdii/cloud-bdii-provider/cloud_bdii/core.py", line 143, in __init__
    super(CloudBDII, self).__init__(opts)
  File "/tmp/bdii/cloud-bdii-provider/cloud_bdii/core.py", line 25, in __init__
    self.dynamic_provider = SUPPORTED_MIDDLEWARE[opts.middleware](opts)
  File "/tmp/bdii/cloud-bdii-provider/cloud_bdii/providers/openstack.py", line 58, in __init__
    cacert=cacert)
  File "/tmp/bdii/local/lib/python2.7/site-packages/novaclient/v2/client.py", line 114, in __init__
    403, _LE("'novaclient.v2.client.Client' is not designed to be "
novaclient.exceptions.Forbidden: 'novaclient.v2.client.Client' is not designed to be initialized directly. It is inner class of novaclient. You should use 'novaclient.client.Client' instead. Related lp bug-report: 1493576 (HTTP 403)
```

File: README.md (hunk 1 above)
Comment (gwarf) [inline README.md]: Done, I updated the repo URL inside the rpmspec and debian/control files too.

**Comment by gwarf** [comment]
Hi @alvarolopez, thanks for looking at this!

Did you test the latest version of the PR? This deprecation warning should already have been fixed by https://github.com/EGI-FCTF/cloud-bdii-provider/pull/30/commits/e147570dbb8367c004626761aabfee0b4b403db8 and https://github.com/EGI-FCTF/cloud-bdii-provider/pull/30/commits/4b1bbf30bc7aa3f0fe77a94d84a1ec6cfae6d8a0 I fixed it for INDIGO, this should have disappeared.

As far as I understand, pbr is only a build dependency and python-novaclient is only needed when using OpenStack, for OpenNebula this is not required. (Are you speaking about requirements.txt or spec/control files?)

**Comment by alvarolopez** [comment]
I am speaking about the requirements.txt file. You are right that "pbr" is required only for building, but IMO it should be present in the "requirements.txt". I do agree with the novaclient requirement, I am too OpenStack-biased :grin:

**Comment by gwarf** [comment]
Here is a commit for this. I also added a commit to disable verbose output of some python dependencies. (verbose output seems to be dependent on the version of the libraries used, but it shouldn't hurt)

**Comment by gwarf** [comment]
Removed the JSON templates for INDIGO.

@cgimeno  @enolfc @alvarolopez : 
There is still an addional OpenNebula provider named IndigoONProvider. It was created to avoid breaking existing behaviour and because in INDIGO we need to have access to the raw images information (without looking at the templates and with real image IDs) and also the raw templates information without using rOCCI.
Despite its name I think it is not very INDIGO specific as it was mostly added to return almost direct infromation from the ON CMF. It won't be used unless explicitely required so I think it is acceptable to have it in the code (especialy as it is required to having the complete/raw CMF data to later do whatever might be needed in the templates), it can perhaps be renamed to something more generic. (or even merged with the existing OpenNebulaProvider as I think that for ON sites they are asked to use the OpenNebulaROCCIProvider?)

**Comment by gwarf** [comment]
I just rebased the PR so that this is no more outdated.

**Comment by enolfc** [comment]
@gwarf 
ONe sites use the OpenNebulaROCCIProvider. I see no problem in leaving the "indigo" provider there. If @cgimeno needs to use it for the new schema it can be renamed then

**Comment by cgimeno** [comment]
There are some minor differences on the new version. For example, the new version outputs the images using the following format:

GLUE2EntityName: resource_tpl#m1-tiny

And the format for the old one is this:

GLUE2EntityName: resource_tpl#aca62593-5556-44e7-bfaa-0b4fe1a03395

Is this the expected behaviour @enolfc  @alvarolopez?

Also @gwarf there is an error on "storage_service.ldif". In line 11 where it reads:

${attributes['storage_service_name}_cloud.storage

Should be replaced by:

${attributes['storage_service_name']}_cloud.storage

**Comment by enolfc** [comment]
> For example, the new version outputs the images using the following format:
>GLUE2EntityName: resource_tpl#m1-tiny
>
>And the format for the old one is this:
>
>GLUE2EntityName: resource_tpl#aca62593-5556-44e7-bfaa-0b4fe1a03395

are you mixing images and flavors here? The GLUE2EntityName should be the id as available in OCCI (depends if you are using ooi or OCCI-OS).

**Comment by cgimeno** [comment]
Ooops, you're right I've been comparing the new output with the old one which is using "--legacy-occi-os" option (yes, we're still using occi-os in our endpoints).

Without this option the result is very similar, only the dn changes for images.

For example, new bdii:

dn: GLUE2ResourceID=aca62593-5556-44e7-bfaa-0b4fe1a03395_https://server4-eupt.unizar.es:5000/v2.0,GLUE2ServiceID=https://server4-eupt.unizar.es:5000/v2.0_cloud.compute,GLUE2GroupID=cloud,o=glue

Old one:

dn: GLUE2ResourceID=resource_tpl#aca62593-5556-44e7-bfaa-0b4fe1a03395_https://server4-eupt.unizar.es:5000/v2.0,GLUE2ServiceID=https://server4-eupt.unizar.es:5000/v2.0_cloud.compute,GLUE2GroupID=cloud,o=glue

As you can see, the old one has "resource_tpl" before the ID.

**Comment by arax** [comment]
Since we are already discussing `GLUE2EntityName`s for resource templates and OS templates, would it be possible to include full URIs instead of just the last path element and fragment? This is crucial because only the full URI is guaranteed to be unique within the given domain (OOI/OCCI-OS/rOCCI-server instance).

**Comment by enolfc** [comment]
@arax: See #15. @cgimeno, please consider this issue also when working on the next features.

**Comment by cgimeno** [comment]
@enolfc Sure, I'll take it into consideration.

I think that this version is a good starting point to implement all the new features.

@enolfc or @alvarolopez could you merge @gwarf pull-request? I don't have enough permissions to do it.

**Comment by gwarf** [comment]
@cgimeno in fact this is a bit strange, the resource_tpl is supposed to be set by the opennebula ROCIC provider and I didn't modify it at all. According to https://github.com/EGI-FCTF/cloud-bdii-provider/blob/master/cloud_bdii/providers/opennebula.py#L175 it looks it would be set from the configuration, using something (uncommented) like this: https://github.com/EGI-FCTF/cloud-bdii-provider/blob/master/etc/sample.opennebularocci.yaml#L72 . Are you using the exact same configuration files?

**Comment by cgimeno** [comment]
Yes, I just copied my current configuration at server4-eupt.unizar.es to my devel machine and after that I changed the name from bdii.yaml to static.yaml.

**Comment by gwarf** [comment]
And does it include the schema: resource_tpl under templates/defaults? If not, does it work when adding it?

**Comment by cgimeno** [comment]
Only this is at this section:

    templates:
        defaults:
            platform: amd64
            network: public

    images:

        defaults:
            # Set to False or comment the line below if you want to show
            # all the images installed in the site (also snapshots). Otherwise
            # only images with a valid marketplace ID (set by the marketplace
            # custom property) are shown
            require_marketplace_id: true
            platform: amd64

The same as the sample.openstack.yaml file

**Comment by cgimeno** [comment]
@gwarf this behaviour is a bug in your ldif files. If you take a look into compute_bdii.ldif line 55 it reads:

dn: GLUE2ResourceID=${template_id}_${template['compute_service_name']},GLUE2ServiceID=${template['compute_service_name']}_cloud.compute,GLUE2GroupID=cloud,${template['suffix']}

And it should be:

dn: GLUE2ResourceID=${template['template_id']}_${template['compute_service_name']},GLUE2ServiceID=${template['compute_service_name']}_cloud.compute,GLUE2GroupID=cloud,${template['suffix']}

**Comment by gwarf** [comment]
Great @cgimeno, you found it! I was wrongly taking granted that the key template_id and the template_id in the dict would stay consistent but it is not the case. Need to learn reading :) (but in fact it is good as it allows to access the real , prefix-less template_id from the template if needed) Your tests and review was definitely useful, thanks and sorry for the error.

**Comment by cgimeno** [comment]
You're welcome :)

The output now is pretty much the same. The new one only has a few cosmetic changes, for example it outputs a brief description of the images under GLUE2ApplicationEnvironmentDescription when the old one only writes "None" on this field

@enolfc or @alvarolopez could you please merge it? :+1:

**Comment by gwarf** [comment]
Great! About the description let's say that "this is a feature not a bug" ©® " 'Thanks for the merge @enolfc it is great to see this be merged!

**Pull request merged**